### PR TITLE
Add delete and add methods for parsers in global configuration

### PR DESCRIPTION
**Pull request by meiswjn**
I have a Jenkins instance that multiple teams work on at the same time. Since it requires admin permissions to add a new parser via the UI, we allow users to add them via Groovy. However, the function `ParserConfiguration#setParsers` is much more prone to accidentally deleting all other parsers.

With this PR, single parsers can be added and deleted from within pipelines.

### Testing done
-> Unit Tests and tested in localhost instance

```[tasklist]
### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

### CODE REVIEW

**Comment by meiswjn** [comment]
I want to implement some logging as well but it is already open for feedback @uhafner

**Comment by meiswjn** [comment]
Thanks for your input @KalleOlaviNiemitalo!

In our case, the methods are called directly from pipelines. We want to allow users to create their own parsers without using setParsers, because it is much more prone to accidentially delete all other parsers.

-> We also don't need DataBoundSetter, thanks!

**Comment by uhafner** [comment]
> I have a Jenkins instance that multiple teams work on at the same time. Since it requires admin permissions to add a new parser via the UI, we allow users to add them via Groovy. However, the function `ParserConfiguration#setParsers` is much more prone to accidentally deleting all other parsers.
> 
> With this PR, single parsers can be added and deleted from within pipelines.

Can you please show how you plan to invoke these methods?

**Comment by meiswjn** [comment]
Sure.

Here is a real example (edited actual values) that I grabbed from a user. In this example, they delete all other parsers. Of course they could simply extend the list, but they have to know that.
```groovy
def InitParser(){
    def config = io.jenkins.plugins.analysis.warnings.groovy.ParserConfiguration.getInstance()
    config.setParsers([]) // Here, the user accidentially deleted all other parsers.

    def newParser = new io.jenkins.plugins.analysis.warnings.groovy.GroovyParser(
        'some_id', 
        'some_name', 
        'my_regex', 
        '''
        ...script...
        ''...',
        "..."
        )
    config.setParsers([newParser]) // Same thing: All other parsers deleted by accident.
}
InitParser()
```
With this PR, you can do (the script that I tested myself with the new changes):
```groovy
def InitParser(){
    def config = io.jenkins.plugins.analysis.warnings.groovy.ParserConfiguration.getInstance()

    def newParser = new io.jenkins.plugins.analysis.warnings.groovy.GroovyParser(
        'some_id', 
        'some_name', 
        'my_regex', 
        '''
        ...script...
        ''...',
        "..."
        )
    config.addParser(newParser)
}
InitParser()
```

**Comment by uhafner** [comment]
> Here is a real example (edited actual values) that I grabbed from a user. In this example, they delete all other parsers. Of course they could simply extend the list, but they have to know that.
> 

Interesting. Which permissions do users have to execute such code that normally requires admin permission?

**Comment by meiswjn** [comment]
No special permission is needed. However, the method must be allowed via Script Approval. Because of this, we would also be able to prevent users from using `#setParsers`

**Comment by uhafner** [comment]
> Because of this, we would also be able to prevent users from using #setParsers and exclusively allow the add, get and delete method.

It definitely is better than allowing them to directly use `set`. But I'm not sure if your approach is a good thing from security perspective: this basically allows your users to completely erase an agent disk without further approval.

**Comment by meiswjn** [comment]
How would they be able to do that with those two methods? From my understanding they are only able to make changes to the `ParserConfiguration` which does not relate to the agent disk.

**Comment by uhafner** [comment]
> > I guess @uhafner means the Groovy code in the parser runs without a Groovy sandbox and has the same access to files as the Jenkins agent process. (I'm not sure whether parsers run in agents or in the Jenkins controller.)
> > https://github.com/jenkinsci/warnings-ng-plugin/blob/610b02f95ebbc257c7cdd071b70d8322cbf89115/plugin/src/main/java/io/jenkins/plugins/analysis/warnings/groovy/GroovyExpressionMatcher.java#L23-L25
> > 
> > Apart from that, I think there might be a conflict if multiple pipelines attempt to register the parser with the same ID in parallel.
> 
> The code runs sandboxed. I was also surprised by this, but it is the case.
> 

The code runs *without* sandboxing on an agent (agents do not use the sandbox, this is a controller concept only). And running Groovy parsers on the controller is not allowed.

**Comment by meiswjn** [comment]
> > The code runs sandboxed. I was also surprised by this, but it is the case.
> 
> I don't see where the sandbox is set up. How did you test it?

In the pipeline itself:
image


> If `setParsers` is only called during Jenkins startup or when an administrator edits the settings, but `addParser` is intended to be called from pipelines, then that increases the conflict chance. If a multibranch pipeline hardcodes the parser ID, and the parser script is edited in one branch but the ID is not changed, then it is not clear which version of the parser will be used.
> 
> If `addParser` were only called from [hook groovy scripts](https://www.jenkins.io/doc/book/managing/groovy-hook-scripts/) and not from pipelines, then that would prevent conflicts but would not let non-administrators add parsers.

You are correct in your scenario, but I am comparing the scenarios where a user calls `setParsers` vs `addParser` directly from the pipeline. In my scenario, the conflict chance is the same for both.

**Comment by uhafner** [comment]
> > > The code runs sandboxed. I was also surprised by this, but it is the case.
> > 
> > I don't see where the sandbox is set up. How did you test it?
> 
> In the pipeline itself: image
> 

This does not affect the Groovy parser, as this parser is invoked on the agent. (See my comment above.)

**Comment by meiswjn** [comment]
> > > I guess @uhafner means the Groovy code in the parser runs without a Groovy sandbox and has the same access to files as the Jenkins agent process. (I'm not sure whether parsers run in agents or in the Jenkins controller.)
> > > https://github.com/jenkinsci/warnings-ng-plugin/blob/610b02f95ebbc257c7cdd071b70d8322cbf89115/plugin/src/main/java/io/jenkins/plugins/analysis/warnings/groovy/GroovyExpressionMatcher.java#L23-L25
> > > 
> > > Apart from that, I think there might be a conflict if multiple pipelines attempt to register the parser with the same ID in parallel.
> > 
> > 
> > The code runs sandboxed. I was also surprised by this, but it is the case.
> 
> The code runs _without_ sandboxing on an agent (agents do not use the sandbox, this is a controller concept only). And running Groovy parsers on the controller is not allowed.

But you would still need to allow the function in the script approval on a per-function basis. From what I understand, the mentioned ParserConfiguration functions are not able to erase an agents disk.

Users are not able to run any other commands except for `#deleteParser`, `#addParser` and `#getParsers`.

**Comment by uhafner** [comment]
> ```groovy
> ...script...
> ```

This is not true. Users can run any groovy script that is passed in the parameter `...script...`. Hopefully this value is hard coded in your pipelines so that you can review it, but you need to be aware that this code can be misused.

**Comment by meiswjn** [comment]
> I wonder if the parsers could be set up as local to the pipeline, without changing the global configuration. Similar to how the `withChecks` step works. Then the parsers would not affect any other Job, and not even any other Run of the current Job.
> 
> ```groovy
> withWarningsGroovyParsers([[
>     id: 'some_id',
>     name: 'some_name',
>     regexp: 'my_regex',
>     script: '...script...', // or perhaps readTrusted('my_parser.groovy') for easier editing
>     example: '...']])
> {
>     // Use the recordIssues step here.
> }
> ```
> 
> > And running Groovy parsers on the controller is not allowed.
> 
> Is that restriction specific to Groovy parsers, or does it just follow from how the `recordIssues` and `scanForIssues` pipeline steps work?

-> But that would not resolve this concern, if I am correct?

> OTOH, if a pipeline can add an unsandboxed Groovy parser to the global configuration and this parser ends up being executed by other pipelines on other agents, then that seems a privilege escalation risk.

**Comment by uhafner** [comment]
> > And running Groovy parsers on the controller is not allowed.
> 
> Is that restriction specific to Groovy parsers, or does it just follow from how the `recordIssues` and `scanForIssues` pipeline steps work?

All parsers run on the agent since they might need access to other local files.

**Comment by uhafner** [comment]
> > I wonder if the parsers could be set up as local to the pipeline, without changing the global configuration. Similar to how the `withChecks` step works. Then the parsers would not affect any other Job, and not even any other Run of the current Job.
> > ```groovy
> > withWarningsGroovyParsers([[
> >     id: 'some_id',
> >     name: 'some_name',
> >     regexp: 'my_regex',
> >     script: '...script...', // or perhaps readTrusted('my_parser.groovy') for easier editing
> >     example: '...']])
> > {
> >     // Use the recordIssues step here.
> > }
> >   
> > > And running Groovy parsers on the controller is not allowed.
> > 
> > Is that restriction specific to Groovy parsers, or does it just follow from how the `recordIssues` and `scanForIssues` pipeline steps work?
> 
> -> But that would not resolve this concern, if I am correct?
> 

BTW: this issue is tracked as https://issues.jenkins.io/browse/JENKINS-52237.

**Comment by uhafner** [comment]
> -> But that would not resolve this concern, if I am correct?

Just to make sure: I have no concern about this PR, it just simplifies something that is already possible (and documented in https://github.com/jenkinsci/warnings-ng-plugin/blob/master/doc/Documentation.md#creating-support-for-a-custom-tool). 

I just wanted to make you aware that this might introduce a security problem.

File: plugin/src/main/java/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration.java [hunk 1]
```diff
@@ -87,6 +87,38 @@ public void setParsers(final List<GroovyParser> parsers) {
...
+     * @param parserId
+     *         the ID of the Groovy parser to be deleted
+     */
+    public void deleteParser(final String parserId) {
+        if (contains(parserId)) {
+            GroovyParser parser = getParser(parserId);
+            this.parsers.remove(parser);
+        }
+        else {
+            throw new NoSuchElementException("No parser with this ID.");
```
Comment (uhafner) [inline plugin/src/main/java/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration.java:102]: Which ID? Please add the ID to the message.

File: plugin/src/main/java/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration.java [hunk 2]
```diff
@@ -87,6 +87,38 @@ public void setParsers(final List<GroovyParser> parsers) {
...
+
+    /**
+     * Adds a GroovyParser to the list without removing other parsers.
+     *
+     * @param parser
+     *         the new Groovy parser to be added
+     */
+    public void addParser(final GroovyParser parser) {
+        if (contains(parser.getId())) {
+            throw new IllegalArgumentException("ID already exists.");
```
Comment (uhafner) [inline plugin/src/main/java/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration.java:115]: ID, see above.

File: plugin/src/test/java/io/jenkins/plugins/analysis/warnings/groovy/ParserConfigurationTest.java [hunk 3]
```diff
@@ -68,4 +68,46 @@ void shouldWarnUserIfConsoleLogScanningPermittedIsSet() {
         final FormValidation actualTrue = configuration.doCheckConsoleLogScanningPermitted(true);
         assertThat(actualTrue.kind).isEqualTo(FormValidation.Kind.WARNING);
     }
+
+    @Test
+    void shouldSaveConfigurationIfParserIsAdded() {
+        GlobalConfigurationFacade facade = mock(GlobalConfigurationFacade.class);
+        GroovyParser test_parser = new GroovyParser("1", "", "", "", "");
```
Comment (uhafner) [inline plugin/src/test/java/io/jenkins/plugins/analysis/warnings/groovy/ParserConfigurationTest.java:75]:
```suggestion
        GroovyParser additionalParser = new GroovyParser("1", "", "", "", "");
```

File: plugin/src/test/java/io/jenkins/plugins/analysis/warnings/groovy/ParserConfigurationTest.java [hunk 4]
```diff
@@ -68,4 +68,46 @@ void shouldWarnUserIfConsoleLogScanningPermittedIsSet() {
...
+
+        ParserConfiguration configuration = new ParserConfiguration(facade);
+        configuration.addParser(test_parser);
+
+        verify(facade).save();
+        assertThat(configuration.getParsers().contains(test_parser));
```
Comment (uhafner) [inline plugin/src/test/java/io/jenkins/plugins/analysis/warnings/groovy/ParserConfigurationTest.java:81]: This does not test anything. 

```suggestion
        assertThat(configuration.getParsers()).containsExactly(test_parser);
```

File: plugin/src/test/java/io/jenkins/plugins/analysis/warnings/groovy/ParserConfigurationTest.java [hunk 5]
```diff
@@ -68,4 +68,46 @@ void shouldWarnUserIfConsoleLogScanningPermittedIsSet() {
...
+    @Test
+    void shouldThrowIfParserExists() {
+        GlobalConfigurationFacade facade = mock(GlobalConfigurationFacade.class);
+        GroovyParser testParser = new GroovyParser("1", "", "", "", "");
+
+        ParserConfiguration configuration = new ParserConfiguration(facade);
+        configuration.addParser(testParser);
+        verify(facade).save();
+
+        assertThatIllegalArgumentException().isThrownBy(() -> configuration.addParser(testParser));
```
Comment (uhafner) [inline plugin/src/test/java/io/jenkins/plugins/analysis/warnings/groovy/ParserConfigurationTest.java:93]: When you add the ID to the message then you can add `.withMessageContaining(ID)`

File: plugin/src/test/java/io/jenkins/plugins/analysis/warnings/groovy/ParserConfigurationTest.java [hunk 6]
```diff
@@ -68,4 +68,46 @@ void shouldWarnUserIfConsoleLogScanningPermittedIsSet() {
...
+    void deleteShouldRemoveOnlySpecifiedParser() {
+        GlobalConfigurationFacade facade = mock(GlobalConfigurationFacade.class);
+        GroovyParser firstTestParser = new GroovyParser("1", "", "", "", "");
+        GroovyParser secondTestParser = new GroovyParser("2", "", "", "", "");
+
+        ParserConfiguration configuration = new ParserConfiguration(facade);
+        configuration.addParser(firstTestParser);
+        configuration.addParser(secondTestParser);
+
+        assertThat(configuration.getParsers().contains(firstTestParser));
```
Comment (uhafner) [inline plugin/src/test/java/io/jenkins/plugins/analysis/warnings/groovy/ParserConfigurationTest.java:106]:
```suggestion
        assertThat(configuration.getParsers()).containsExactly(firstTestParser, secondParser);
```

File: plugin/src/test/java/io/jenkins/plugins/analysis/warnings/groovy/ParserConfigurationTest.java [hunk 7]
```diff
@@ -68,4 +68,46 @@ void shouldWarnUserIfConsoleLogScanningPermittedIsSet() {
...
+        assertThat(configuration.getParsers().contains(secondTestParser));
+
+        configuration.deleteParser(firstTestParser.getId());
+        assertThat(!(configuration.getParsers().contains(firstTestParser)));
+        assertThat(configuration.getParsers().contains(firstTestParser));
```
Comment (uhafner) [inline plugin/src/test/java/io/jenkins/plugins/analysis/warnings/groovy/ParserConfigurationTest.java:111]:
```suggestion
    assertThat(configuration.getParsers()).containsExactly(secondParser);
```

**Review by uhafner — CHANGES_REQUESTED** [review]
As @KalleOlaviNiemitalo mentioned, your code is not thread safe and might produce problems in the way you are intending to use it. But I think we can ignore that right now and change that later if users complain.

Optional enhancement: add an integration test that shows your approach in code, example https://github.com/jenkinsci/warnings-ng-plugin/blob/51cca6b65de2e22d65f2bb16f8dd80a10aa11e20/plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsITest.java#L799.

**Comment by meiswjn** [comment]
Thanks for your review, @uhafner. I implemented the changes and one integration test.

**Review by uhafner — APPROVED** [review]
Thanks!

**Pull request merged**